homeassistant/components/device_tracker/icloud.py

```diff
@@ -1,100 +1,455 @@
 """
-Support for iCloud connected devices.
+Platform that supports scanning iCloud.
 
 For more details about this platform, please refer to the documentation at
 https://home-assistant.io/components/device_tracker.icloud/
 """
 import logging
+import random
+import os
+
 import voluptuous as vol
 
-from homeassistant.const import (CONF_PASSWORD, CONF_USERNAME,
-                                 EVENT_HOMEASSISTANT_START)
+from homeassistant.const import CONF_USERNAME, CONF_PASSWORD
+from homeassistant.components.device_tracker import (
+    PLATFORM_SCHEMA, DOMAIN, ATTR_ATTRIBUTES, ENTITY_ID_FORMAT)
+from homeassistant.components.zone import active_zone
 from homeassistant.helpers.event import track_utc_time_change
+import homeassistant.helpers.config_validation as cv
 from homeassistant.util import slugify
-from homeassistant.components.device_tracker import (ENTITY_ID_FORMAT,
-                                                     PLATFORM_SCHEMA)
+import homeassistant.util.dt as dt_util
+from homeassistant.util.location import distance
+from homeassistant.loader import get_component
 
 _LOGGER = logging.getLogger(__name__)
 
 REQUIREMENTS = ['pyicloud==0.9.1']
 
-CONF_INTERVAL = 'interval'
-KEEPALIVE_INTERVAL = 4
+CONF_IGNORED_DEVICES = 'ignored_devices'
+CONF_GMTT = 'google_maps_travel_time'
+CONF_COOKIEDIRECTORY = 'cookiedirectory'
+CONF_ACCOUNTNAME = 'account_name'
+
+# entity attributes
+ATTR_ACCOUNTNAME = 'account_name'
+ATTR_INTERVAL = 'interval'
+ATTR_DEVICENAME = 'device_name'
+ATTR_BATTERY = 'battery'
+ATTR_DISTANCE = 'distance'
+ATTR_DEVICESTATUS = 'device_status'
+ATTR_LOWPOWERMODE = 'low_power_mode'
+ATTR_BATTERYSTATUS = 'battery_status'
+ATTR_GMTT = 'google_maps_travel_time'
+ATTR_GMTT_DURATION = 'gmtt_duration'
+ATTR_GMTT_ORIGIN = 'gmtt_origin'
+
+ICLOUDTRACKERS = {}
+
+_CONFIGURING = {}
+
+DEVICESTATUSSET = ['features', 'maxMsgChar', 'darkWake', 'fmlyShare',
+                   'deviceStatus', 'remoteLock', 'activationLocked',
+                   'deviceClass', 'id', 'deviceModel', 'rawDeviceModel',
+                   'passcodeLength', 'canWipeAfterLock', 'trackingInfo',
+                   'location', 'msg', 'batteryLevel', 'remoteWipe',
+                   'thisDevice', 'snd', 'prsId', 'wipeInProgress',
+                   'lowPowerMode', 'lostModeEnabled', 'isLocating',
+                   'lostModeCapable', 'mesg', 'name', 'batteryStatus',
+                   'lockedTimestamp', 'lostTimestamp', 'locationCapable',
+                   'deviceDisplayName', 'lostDevice', 'deviceColor',
+                   'wipedTimestamp', 'modelDisplayName', 'locationEnabled',
+                   'isMac', 'locFoundEnabled']
+
+DEVICESTATUSCODES = {'200': 'online', '201': 'offline', '203': 'pending',
+                     '204': 'unregistered'}
+
+SERVICE_SCHEMA = vol.Schema({
+    vol.Optional(ATTR_ACCOUNTNAME): vol.All(cv.ensure_list, [cv.string]),
+    vol.Optional(ATTR_DEVICENAME): cv.string,
```

Comment:
Should devicename not be a `cv.slug` or `cv.entity_id`? And do you need a domain? `device_tracker.iphone`?